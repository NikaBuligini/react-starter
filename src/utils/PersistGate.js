/* eslint-disable no-underscore-dangle */
// @flow

import React from 'react';
import type { Node } from 'react';

type Props = {
  onBeforeLift?: Function,
  children: Node,
  loading?: Node,
  persistor: Object,
  ignoreBootstrap?: boolean,
};

type State = {
  bootstrapped: boolean,
};

class PersistGate extends React.PureComponent<Props, State> {
  static defaultProps = {
    loading: null,
    ignoreBootstrap: false,
    onBeforeLift: undefined,
  };

  state = {
    bootstrapped: false,
  };

  componentDidMount() {
    this._unsubscribe = this.props.persistor.subscribe(this.handlePersistorState);
    this.handlePersistorState();
  }

  componentWillUnmount() {
    if (this._unsubscribe) {
      this._unsubscribe();
    }
  }

  handlePersistorState = () => {
    const { persistor, onBeforeLift } = this.props;
    const { bootstrapped } = persistor.getState();
    if (bootstrapped) {
      if (onBeforeLift) {
        Promise.resolve(onBeforeLift())
          .then(() => this.setState({ bootstrapped: true }))
          .catch(() => this.setState({ bootstrapped: true }));
      } else {
        this.setState({ bootstrapped: true });
      }

      if (this._unsubscribe) {
        this._unsubscribe();
      }
    }
  };

  _unsubscribe: ?Function;

  render() {
    const { bootstrapped } = this.state;
    const { ignoreBootstrap, loading, children } = this.props;

    if (ignoreBootstrap) {
      return children;
    }

    return bootstrapped ? children : loading;
  }
}

export default PersistGate;
