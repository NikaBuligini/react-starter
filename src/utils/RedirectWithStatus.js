// @flow

import React from 'react';
import { Route, Redirect } from 'react-router-dom';
// import type { ContextRouter } from 'react-router-dom';

const RedirectWithStatus = ({ from, to, status }: { from: string, to: string, status: number }) => (
  <Route
    render={(router: Object) => {
      // router actually is type of ContextRouter
      const { staticContext } = router;

      // there is no `staticContext` on the client, so
      // we need to guard against that here
      if (staticContext) {
        staticContext.status = status;
      }

      return <Redirect from={from} to={to} />;
    }}
  />
);

/**
 * Example:
 * -------------------
 * <RedirectWithStatus
 *   status={301}
 *   from="/"
 *   to="/auth/signIn"
 * />
 */

export default RedirectWithStatus;
