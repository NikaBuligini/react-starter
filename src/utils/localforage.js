import localforage from 'localforage';

localforage.config({
  driver: [localforage.WEBSQL, localforage.INDEXEDDB, localforage.LOCALSTORAGE],
  name: 'react-starter',
  version: 1.0,
  size: 4980736, // Size of database, in bytes. WebSQL-only for now.
  storeName: 'react_starter_app', // Should be alphanumeric, with underscores.
  description: 'Local storage database for react app',
});

export default localforage;
