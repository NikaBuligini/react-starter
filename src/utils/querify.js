// @flow

type QueryParams = {
  [key: string]: string | number,
};

export default function querify(endpoint: string, queryParams: QueryParams) {
  const query = Object.keys(queryParams)
    .map(key => `${key}=${queryParams[key]}`)
    .join('&');

  return `${endpoint}?${query}`;
}
