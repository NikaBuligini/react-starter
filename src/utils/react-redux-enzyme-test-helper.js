/**
 * Components using the react-redux module require access to the redux context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap with a valid store.
 */

import React from 'react';
import { applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

import configureStore from '../store/configureStore';

import api from '../middlewares/api';
import array from '../middlewares/array';
import debounce from '../middlewares/debounce';
import analytics from '../middlewares/analytics';
import reducers from '../reducers';

const store = configureStore(
  undefined,
  applyMiddleware(thunk, array, api, debounce, analytics),
  reducers,
);

/* eslint-disable import/prefer-default-export */

export function wrapWithProvider(component) {
  return <Provider store={store}>{React.cloneElement(component)}</Provider>;
}
