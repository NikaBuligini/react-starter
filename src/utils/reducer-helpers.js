// @flow
/* eslint-disable import/prefer-default-export */

export function isFullyLoaded(action: Object, ids?: Array<number>) {
  const data = ids || action.response.result.data;

  return typeof action.limit === 'number' ? data.length < action.limit : false;
}
