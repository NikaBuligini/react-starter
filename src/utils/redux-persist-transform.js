import { createTransform } from 'redux-persist';

/**
 * specialSerialize
 * @param state
 * @param key
 */
function serialize(state) {
  return state;
}

/**
 * specialDeserialize
 * @param state
 * @param key
 */
function deserialize(state) {
  return state;
}

const transform = createTransform(
  // transform state coming from redux on its way to being serialized and stored
  (state, key) => serialize(state, key),
  // transform state coming from storage, on its way to be rehydrated into redux
  (state, key) => deserialize(state, key),
  // configuration options
  { whitelist: ['session'] },
);

export default transform;
