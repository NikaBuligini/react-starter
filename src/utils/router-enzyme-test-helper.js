/**
 * Components using the react-router module require access to the Router context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap with a valid MemoryRouter.
 */

import React from 'react';
import { MemoryRouter } from 'react-router-dom';

/* eslint-disable import/prefer-default-export */

export function wrapWithRouter(component, routerProps = {}) {
  return <MemoryRouter {...routerProps}>{React.cloneElement(component)}</MemoryRouter>;
}
