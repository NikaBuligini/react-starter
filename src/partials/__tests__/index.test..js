import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { wrapWithProvider } from '../../utils/redux-enzyme-test-helper';
import { wrapWithRouter } from '../../utils/router-enzyme-test-helper';
import { Footer, Navigation } from '../index';

describe('<Partials />', () => {
  it('should match Footer snapshot', () => {
    const component = renderer.create(wrapWithProvider(<Footer />));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have select node', () => {
    const renderedComponent = mount(wrapWithProvider(<Footer />));
    expect(renderedComponent.find('.language-switch').length).toBe(1);
  });

  it('should match Navigation snapshot', () => {
    const component = renderer.create(wrapWithRouter(<Navigation />));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match Navigation snapshot when route is /contributors', () => {
    const component = renderer.create(
      wrapWithRouter(<Navigation />, { initialEntries: ['/contributors'] }),
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
