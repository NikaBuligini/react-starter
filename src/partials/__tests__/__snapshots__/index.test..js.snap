// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Partials /> should match Footer snapshot 1`] = `
<div
  className="s11ntyxv-0 bxWkpQ"
>
  <div
    className="footer__inner"
  >
    Â©Example
    <select
      className="language-switch"
      onChange={[Function]}
      value="en"
    >
      <option
        value="en"
      >
        en
      </option>
      <option
        value="ka"
      >
        ka
      </option>
    </select>
  </div>
</div>
`;

exports[`<Partials /> should match Navigation snapshot 1`] = `
<div
  className="rahdj5-0 loBpXO"
>
  <a
    aria-current="true"
    className="rahdj5-1 fvaGDL active"
    href="/"
    onClick={[Function]}
    style={Object {}}
  >
    Home
  </a>
  <a
    aria-current={false}
    className="rahdj5-1 fvaGDL"
    href="/contributors"
    onClick={[Function]}
    style={undefined}
  >
    Contributors
  </a>
</div>
`;

exports[`<Partials /> should match Navigation snapshot when route is /contributors 1`] = `
<div
  className="rahdj5-0 loBpXO"
>
  <a
    aria-current={false}
    className="rahdj5-1 fvaGDL"
    href="/"
    onClick={[Function]}
    style={undefined}
  >
    Home
  </a>
  <a
    aria-current="true"
    className="rahdj5-1 fvaGDL active"
    href="/contributors"
    onClick={[Function]}
    style={Object {}}
  >
    Contributors
  </a>
</div>
`;
